### ISSUE TITLE
UI Testing of Automation Conditions Does Not Work with Template Conditions

### ISSUE THREAD

**Issue by marlemion** [home-assistant/frontend#12282]
### The problem

In order to implement a somewhat more complex template condition, I came across the problem that (at least testing) templates conditions do not work (anymore).

What I did:

Create a simple automation having a template condition:

```
  - condition: template
    value_template: "{{ 1 > 2 }}"

```
So, this is extremely simplifed. The expression should return False, but instead I get the following error when testing this expression in the condition frontend (under automatisations):

`Error handling message: template value should be a string for dictionary value @ data['value_template']. Got None (invalid_format)`

If I enter the same snippet of code in the templates section un der developer tools, it prints false.

Either there is a problem in the code just adding the template condition in the automatisation frontend or something is completely borked.

### What version of Home Assistant Core has the issue?

2022.4.1

### What type of installation are you running?

Home Assistant Container

### Integration causing the issue

Automation

### Example YAML snippet

```yaml
- condition: template
    value_template: "{{ 1 > 2 }}"
```

### Anything in the logs that might be useful for us?

```txt
2022-04-08 14:40:54 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140209354674720] Error handling message: template value should belue_template']. Got None (invalid_format)
```

**Comment by Plawasan** [comment]
Just to be clear, in my experience it's only the actual test button that doesn't show the correct result, the template itself and the automation condition still work as expected.

**Comment by wwsmithobut** [comment]
I am new to Home Assistant, but I am having what I believe to be the same issue. I am attempting to trigger off the following value_template “{{ is_state_attr('fan.fanlinc_51_d1_37_fan', 'percentage', 33) }}” . The trigger never runs. But it returns the correct true or false in the developer tools, and it gets this error message from test if I try to use it as a condition instead of a trigger: "template value should be a string for dictionary value @ data['value_template']. Got None". So the difference for me from what Plawasan commented is that the automation trigger as I would expect it to, and if I trigger on the device state change, and the template as a condition, it appears that the condition is not satisfied and the automation doesn't trigger.  This is 2022.4.5 running in a VM using VMWare Player 16.
<body>


Version | core-2022.4.5
-- | --
Installation Type | Home Assistant OS


</body>
</html>

**Comment by analogmonster** [comment]
I am having exactly this issue with automations that worked in earlier versions (although I cannot tell you what version this started showing up).  These automations show up in the log file as an Error evaluating condition, same with using the test buttons although conditions are satisfied with developer mode.  

The conditions take the form of:

{{ ( as_timestamp(now()) - as_timestamp(states.automation.***automation_name***.attributes.last_triggered) |int(0) ) > 86400 }}

let me know if there is anything useful I can add.

**Comment by JayRHa** [comment]
I have the same issue:

condition: template
value_template: '{{is_state (''input_boolean.watering'',''on'')}}'

**Comment by ErnestStaug** [comment]
Same here, was thinking that my template doesn't work but it seems like it's a bug in UI

**Comment by pickonedev** [comment]
Same

**Comment by Bert-R** [comment]
I'm getting that same error on this condition:
```
{{ (states.sensor.wk_temperatuur_vloerverwarming_cv_in|float(999.0) - states.sensor.wk_temperatuur_vloerverwarming_vloer_uit|float(0.0)) <= 1.0 }}
```

**Comment by pickonedev** [comment]
Don't use the test "feature", go to Template and test it there, better this way, till they will fix it...

**Comment by kjt407** [comment]
I have the same issue

ha 2022.4.5 on Ubuntu 20.04.2 LTS

**Comment by kjt407** [comment]
> I have the same issue
> 
> ha 2022.4.5 on Ubuntu 20.04.2 LTS



> I have the same issue
> 
> ha 2022.4.5 on Ubuntu 20.04.2 LTS

Currently updated to 2022.4.7
the problem is still

**Comment by encor50** [comment]
Same issue 2022.4.7 on Hyper-V

**Comment by Stefan-OJ** [comment]
I'm having the same issue

![image](https://user-images.githubusercontent.com/52597554/166651227-26ec07e5-4a13-4f72-9471-adca0733b4b1.png)

**Comment by gfrancesco** [comment]
Same here, in my case the error blocks the condition of the automation, so the automation does not start when it should.
It was working fine in previous HA versions, so I can confirm it's a bug introduced in the last couple of months maybe.

**Comment by Tillsy** [comment]
Oh my gosh I've been wracking my brains over this for past hour until I stumbled upon this bug report - didn't realise but YES what I've been dong IS working, it's just the editor that's popping up the error.

**Comment by antoweb** [comment]
Same problem here this template in developer rools works good, but in automation returtn an error:
{{ states.sensor.uptime_raspberry_tv.state < 30 }}
The error is
template value should be a string for dictionary value @ data['value_template']. Got None

**Comment by TheFes** [comment]
> Same here, in my case the error blocks the condition of the automation, so the automation does not start when it should.
> It was working fine in previous HA versions, so I can confirm it's a bug introduced in the last couple of months maybe.

That will be something else, because the templates regarding this issue will work fine in the automation. It is just the test button which throws an error.
Put your template in developer tools > templates and check if it works there

**Comment by Nick4-1** [comment]
I have this problem too.
Is there anything being done about it?

**Comment by lorenzo-deluca** [comment]
Same problem on 2022.6.1

**Comment by Guenni75** [comment]
Same Problem!
Home Assistant Core 2022.6.1
Home Assistant Supervisor 2022.05.3
Home Assistant OS 8.1

**Comment by struikc** [comment]
Same Problem, unfortunately.
Home Assistant Core 2022.6.4

**Comment by petebanham** [comment]
I am having the same issue however my automation does not run. The template statement works in Dev tools but not in the automation

**Comment by Nick4-1** [comment]
@petebanham: my automation also stopped working.
Have a look at https://community.home-assistant.io/t/automation-template-value-should-be-a-string-for-dictionary-value-data-value-template-got-none/419519

**Comment by markdepalma** [comment]
Having same issue still on 2022.6.5. My automation does not trigger, but the template checks out fine in dev tools.

**Comment by isewise** [comment]
I have a simple template to evaluate the battery level of my Shark vacuum. 

`{{state_attr('vacuum.vicky', 'battery_level') > 85 }}`

In the Dev tools -> Template I get a true or false response as expected. In an automation window get the error:

> Error occurred while testing condition
> template value should be a string for dictionary value @ data['value_template']. Got None

**Comment by JohanViljoen** [comment]
I too ran into the same issue (HA 2022.6.3), and have made an observation which _may_ be useful to those with the necessary insight into the inner workings of the automation function. I have a temperature sensor on my water heater, and control over the element, and I also set up an input_number slider to control the target temperature. Switching off uses the following template in the trigger section:
```
{{
  states('sensor.geyser_temperature')|float >=
  states('input_number.geyser_target_temp')|float
}}
```
I.e. trigger when the actual temperature equals or exceeds the target. And this works just fine, both simulated in Dev Tools, and in operational practice.

**_HOWEVER_**

The element is switched **on** at a specific time (and/or when the solar panels are producing nicely), but we need to check that the water is actually cold before doing so: hence we take that very same code above, just changing the comparison to a less-than <, and put it in the _Conditions_ section, instead of in _Triggers,_ and immediately we get the same error:

> Error occurred while testing condition
> template value should be a string for dictionary value @ data['value_template']. Got None

**So it seems to work as a Trigger, but not as a Condition.**
(Also, less importantly, the _Test_ function is available right in the Automation UI editor for template code added under _Conditions_, but not under _Triggers._ To this untrained eye the two seem rather similar, so it was natural to wonder if there is maybe some structural variation under the hood necessitating this difference, and if this perhaps also pertains to the difference in behaviour that is observed.)

With much respect and gratitude to those who labour on behalf of us unwashed masses, to make the world a better, or at least more automated and entertaining place.

**Comment by TheFes** [comment]
@JohanViljoen it will also work as a condition, it is just the check button which will display the error, the condition will actually work fine.

For testing triggers in the GUI it works different, as a trigger is just a moment where something turns from `false` to `true` it makes no sense to have a similar button as on a condition. When the trigger occurs, the GUI will show a green bar which you can click to show the details.

**Comment by tomkennedy513** [comment]
Same problem here: core version 2022.6.6

**Comment by ktown007** [comment]
On 2022.6.6, the template conditions or numeric_state with value_template do to stop automation from triggering. They work in dev tools and give error when click test in UI. `template value should be a string for dictionary value @ data['value_template']. Got None`
```
    condition: template
    value_template: "{{ 1 > 2 }}"
```
or my real goal to debounce a button. 
```
    condition: template
    value_template: "{{ ((as_timestamp(now()) - as_timestamp(state_attr('automation.test_button','last_triggered'))) | int) > 5 }}"
```

**Comment by githubrdonefun** [comment]
Happening for me on 2022.6.7 as well.  Checked in development tools after finding out this was an issue, and it worked fine.

**Comment by stefanodelbosco** [comment]
There is any workaround for this issue? 
For the moment, I check the 'condiction' with  if/else in the Action section..
Anyone have a 'better' idea? 😄

**Comment by maxlin1** [comment]
I have the same Issue {{ (states.sensor.terrase_thermomether_actual_temperature.state | float )  >=
        state_attr('climate.wohnzimmer_thermostat', 'current_temperature' ) }} doesn't work

**Comment by Santanachia** [comment]
@maxlin1 only the test button does not work. if you go to `/developer-tools/template`, there you can check if your code is correct. Automation will work even if the test button does not work.

**Comment by ktown007** [comment]
> @maxlin1 only the test button does not work. if you go to `/developer-tools/template`, there you can check if your code is correct. Automation will work even if the test button does not work.

@Santanachia it is not just the Condition "TEST" button that does not work. Also when I click "RUN ACTIONS" to test automation the Conditions do not work(in my case no debounce).   When triggered from zigbee button it did apply the condition and debounce.

**Comment by Santanachia** [comment]
@ktown007 In my case "run actions" works, even though "test" returns an error. Maybe in your case "test" returns an error because you have an error, not because "test" is broken?

**Comment by ktown007** [comment]
> @ktown007 In my case "run actions" works, even though "test" returns an error. Maybe in your case "test" returns an error because you have an error, not because "test" is broken?

@Santanachia the "RUN ACTIONS" button ignores(or condition does not work) the condition. For example I creates a template with this condition and when I click run actions it still turns on the light.   

    condition: template
    value_template: '{{ 1 > 2 }}'

In my case the debounce template worked in dev tools and physical zigbee button. I was banged my head on the wall many times because the "TEST" and "RUN ACTIONS" did not work. I was thinking the run actions would simulate the full automation, maybe like the name suggests it only does the Actions and I have to run up the two stories and press the real button to test it...

**Comment by aderusha** [comment]
Just got bit by this on 2022.7.2 after spinning a lot of cycles trying to figure out what I’m doing wrong. This is a pretty bad UI situation here, if you’re going to offer a TEST button, it’s probably better if it returns valid results.

I’d posit that returning invalid results is worse than not offering a test function at all.

**Comment by maxlin1** [comment]
@Santanachia in developer-tools/template the Condition 
{{ (states.sensor.terrase_thermomether_actual_temperature.state | float ) >=
state_attr('climate.wohnzimmer_thermostat', 'current_temperature' ) }}

work right True and  False.  Burt in automations it doesn't work

**Comment by maxlin1** [comment]
@[Santanachia] hier die Screenshots
<img width="1083" alt="Bildschirmfoto 2022-07-11 um 08 02 28" src="https://user-images.githubusercontent.com/10881081/178199021-9cb71ffa-ff8d-4616-849d-00841d16a4d4.png">
<img width="737" alt="Bildschirmfoto 2022-07-11 um 08 03 33" src="https://user-images.githubusercontent.com/10881081/178199038-45731f36-658b-4508-9fd1-c4d7628d2ab8.png">

**Comment by Santanachia** [comment]
@maxlin1 
1. You should use `states("sensor.name")` instead of `states.sensor.name.state`
2. The `TEST` button does not work, but you can still save the automation and it will work properly.

**Comment by maxlin1** [comment]
@Santanachia can you help me for the Right Condition ?
<img width="1188" alt="Bildschirmfoto 2022-07-12 um 09 11 22" src="https://user-images.githubusercontent.com/10881081/178430594-d30acd71-a359-45ad-9d42-433f1647b89c.png">

**Comment by maxlin1** [comment]
OH  yes i See Thank You for everything 

This works:
{{ (states("sensor.terrase_thermomether_actual_temperature") | float ) >
state_attr('climate.wohnzimmer_thermostat', 'current_temperature' ) }}

 with core-2022.7.5 version 

You don't can Test ist but it works

<img width="737" alt="Bildschirmfoto 2022-07-11 um 08 03 33" src="https://user-images.githubusercontent.com/10881081/179349536-144dd65b-b053-4329-81d5-b80a05aed79f.png">

**Comment by m1lkman** [comment]
I can't get this to work for me at all with core-2022.7.5. Errors in the UI as described but also the node doesn't execute according to the trace. I get a boolean result in dev tools template but the logic doesn't work as a condition. What am I doing wrong?

```
condition:
  - condition: template
    value_template: >-
      {{ states('sensor.garage_sensor_temperature')|float > states('sensor.backyard_sensor_temperature')|float }}
```

**Comment by maxlin1** [comment]
Hi @m1lkman check if you need
 {{ (states("sensor. **xxxx** ") | float ) }}
{{ state_attr('climate. xxx', ' xx' ) }}


This works for me. But you can not test it because otherwise the error message comes. So I always have to wait until the 
status is reached. :-(

`alias: Temperatur sinkt Fenster öffnen!
description: ''
trigger:
  - platform: state
    entity_id:
      - sensor.terrase_thermomether_actual_temperature
condition:
  - condition: time
    after: '16:00:00'
    before: '23:59:00'
action:
  - if:
      - condition: template
        value_template: >-
          {{ (states("sensor.terrase_thermomether_actual_temperature") | float )
          <
          state_attr('climate.wohnzimmer_thermostat', 'current_temperature' ) }}
      - type: is_not_open
        condition: device
        device_id: xxxxx
        entity_id: binary_sensor.xxx_fenster_state
        domain: binary_sensor
      - type: is_not_open
        condition: device
        device_id: xxxx
        entity_id: binary_sensor.xxxx_fenster_state
        domain: binary_sensor
    then:
      - service: notify.xxx
        data:
          title: 🪟 Fenster aufmachen zum lüften
          message: >-
            {{states("sensor.terrase_thermomether_actual_temperature") }}C° auf
            der Terrrasse und {{ state_attr('climate.wohnzimmer_thermostat',
            'current_temperature' ) }}C° im Wohnzimmer.
mode: single
`

In my experience there are 2 ways to read the sensors.

**Comment by Eckart0** [comment]
Same problem with HA  2022.7.6.
'''
      - condition: template
        value_template: >-
          {{ (state_attr('binary_sensor.sun_on_window','sun_on_window_r33')|int)
          == 0 }}
'''
works in the template editor, but don't work in the automation UI.
Regards, Eckart

**Comment by rlippmann** [comment]
I’ve run into this issue on straight numeric state conditions. 

For some reason the UI adds a value_template: |+ to the condition. This causes the ‘string for dictionary vale error’ as well as causing the condition to fail when the automation is triggered. 

The UI also doesn’t populate the value template field, so it looks like there’s nothing there.

If you go to the yaml editor and remove the value_template line, everything works.

**Comment by m1lkman** [comment]
@rlippmann do you have an example of what works for you?

**Comment by rlippmann** [comment]
![IMG_0118](https://user-images.githubusercontent.com/70883373/181366332-9bb8c077-fabc-431e-9f6d-7357f77fa16a.jpeg)

If you set that up in the editor, then hit edit yaml, you’ll see a value_template: |+ added. Which is what’s causing the editor to throw the dictionary error.

Delete the value_template line from the yaml and the test works. The value_template also causes the automation condition to always fail.

**Comment by ffblanch** [comment]
Same here i hope this will be fixed soon ...

**Comment by gregschwartz** [comment]
Running into this as well. Any idea on a fix?

**Comment by mainmind83** [comment]
Hitting this issue on 2022.7.7

**Comment by Lafunamor** [comment]
hitting this on as well. Thanks to all the hints towards the dev tool templating engine. I was able to use it to verify my automation.

**Comment by tombobs** [comment]
Still here in Frontend 20220802.0 - is anyone working on a fix?

**Comment by han-fastolfe** [comment]
Got this one, too, on latest HA docker.

**Comment by NODeeJay** [comment]
While I am also unhappy that it is not fixed yet, and [I like many of you stepped into the same trap](https://github.com/home-assistant/frontend/issues/12282#issuecomment-1163195702) (caused by frustration, the expectation that more posts in a thread would raise more awareness and many other reasons behind mankind's psychology) since @rlippmann lasts post, which added value to find a solution, the only posts are something like "#MeToo" and to be honest, they make it pretty complicated to find the posts which help developing a solution. 

[🤚 I have the same issue! 👇 Read below](https://github.com/home-assistant/core/issues/74564#issuecomment-1177576245)

Hi there! Sorry to read you are experiencing the same issue as more people here. Nobody likes having issues 😞

Finding an issue on a GitHub issue tracker that matches your problem is kinda nice: At least you know you are not alone. So, let's leave a commit with: "Yeah, I have the same issue", or "+1!!!". 🚫 No!
Please do not create "I have the same issue" comments!

Not just this issue or this project, but anywhere on GitHub or any issue tracker on the internet even. It doesn't add to triaging the issue. It only generates noise when reading / triaging the case.
What to do instead?

Instead, go to the top of the issue, and add a 👍 emoji reaction. That way, we can still see that you (and how many others) experience this issue while keeping the issue threads nice and clean.

Thanks! ❤️

../Frenck

---

@rlippmann I tried your suggestion and added different conditions, however I was not able to reproduce your findings, neither by clicking edit in Yaml at condition level, nor at automation level. I also tried with a value template.
From what I understood, you do not set any value template but use an entity that stores the temperature or the compared value in an attribute.

You choose Numeric State as condition, choose an entity, choose an attribute and define either above/below?

What I have in case of value templates in conditions is the `value_template: >-` but with a minus to remove white lines, not with a plus, how exactly do you get there?

**Comment by malosaa** [comment]
i clicked on the emoji above the post aswell.

But anyway, my config looks like this and doesnt work either.
but the automation works normally.

action:
  - choose:
      - conditions:
          - condition: template
            value_template: "{{ now().month in [1,2,3,4,10,11,12] }}"
          - condition: numeric_state
            entity_id: sensor.gas_usage_day
            above: "4"

**Comment by cbhiii** [comment]
_Yelling into the void as well here_

Trying to use
```
  - condition: state
    entity_id: binary_sensor.bedroom_motion_detect
    state: "on"
    for: "00:{{ states('input_number.bedroom_light_motion_timeout') | int(default=10) }}:00"
```
as well as a condition during an action.

Error message:
```
Invalid config for [automation]: offset 00:{{ states('input_number.bedroom_light_motion_timeout') | int }}:00 should be format 'HH:MM', 'HH:MM:SS' or 'HH:MM:SS.F' for dictionary value @ data['action'][0]['choose'][0]['conditions'][0]['conditions'][2]['for']. Got None. (See ?, line ?).
```

Any update on when this issue might be tackled next? Thank you.

**Comment by malosaa** [comment]
I still don't get why its not fixed yet.

![image](https://user-images.githubusercontent.com/13116501/190143807-b01de37d-6684-4773-b7f1-2c9fdc981f14.png)

If i check the fix from WALSH, it should work, but i did not test yet.
But has been closed for the need of more info / questions what i don't understand...

With my example it should return, true or false, but give an error instead.
I know you can use **developer tools** > **template**.

Then the test button will be useless in the template/conditions section of the automation.


I love Home Assistant, but this is something little what could have been fixed a while ago.

**Comment by d-walsh** [comment]
My PR fixes this issue.  @MartinHjelmare wants further information to merge it.

**Comment by aderusha** [comment]
It seems the current take is "we don't understand the problem so we won't accept any fixes" which is pretty frustrating.  I don't hold out much hope of this bug ever being resolved if fixes won't be accepted.

**Comment by frenck** [comment]
> It seems the current take is "we don't understand the problem so we won't accept any fixes" which is pretty frustrating. 

I'm not sure how it is frustrating. Providing an explanation for what one is changing, why and how one came to this solution... is pretty common, normal and fair IMHO.

>  I don't hold out much hope of this bug ever being resolved if fixes won't be accepted.

? Maybe consider changing your attitude a tiny bit. Everybody is trying to do their best. We all have the same goals. That including making the right fixes.

../Frenck

**Comment by morremeyer** [comment]
Thank you @frenck for taking the time to look into it and get to the bottom of the issue. 

Much appreciated!

**Comment by Santanachia** [comment]
> ? Maybe consider changing your attitude a tiny bit.
@frenck, You should try this too. This is not the only issue in which you block the implementation of a fix.

**Comment by morremeyer** [comment]
While I'm not familiar with other issues, claiming that @frenck is blocking the implementation of a fix here is… wow.

He not only analyzed the proposed fix and found out that it would work around the root cause, he also implemented a fix for the root cause.

Calling that „you block the implementation“ is more than just a stretch, it's an outright lie.

**Comment by Santanachia** [comment]
I am sure that if you ask him about the fix for Tuya integration, he will write you something more. The fix (rollback of changes) was ready as soon as the bug was introduced in the production core without being presented in the beta version first. @frenck blocked the rollback of the changes that caused the bug for several months.

**Comment by Santanachia** [comment]
it's really cool that he fixed the source of the problem. I admire him for covering the whole thing and can't imagine HA without his contribution.

Nevertheless, when there is a fix and there is confirmation that it works, it needs to be implemented without delay.

**Comment by frenck** [comment]
Totally off-topic, but I do know the source of the issue you are referring to. 

Tuya is a mess (upstream) and we have to deal with that (which is hard enough already). Yes, we will hold on to specifications provided by Tuya developers, as there seems to be not a single correct answer to fix all.

Rolling back or not, would in both cases causes issues for a group of people. Meaning there was no single right call. Awaiting adjustment and conclusions from upstream was therefore the only right solution.

If you think I blocked it to pester, you have gotten the wrong impression. If you have been affected, sorry to hear that too. But to be clear Tuya is not our mess.

TL;DR: I (and we as a project) will never block any good PR. If a PR can be properly explained/motivated, brings in a new feature, fix, whatever and fits/adheres to the projects architecture, nothing will stop it.

Workarounds and/or lack of explanation as the original PR for this specific issue had, are therefore declined (for good reason, as the root cause wasn't fixed).

Lastly, to be clear, and not that it matter, in this case I wasn't the one blocking.

../Frenck
